File: sudoku.py
z = 9 #your signature

grid = [[0 for i in range(z)] for j in range(z)]
with open("sudoku.txt", "r") as f:
    c = 0
    for i in range(z):
        row = f.readline().split()
        for j in range(z):
            grid[i][j] = int(row[j])
            if grid[i][j] == 0:
                c += 1


grid_2 = []
for j in range(9):
    row = []
    for i in range(9):
        row.append(grid[j][i])
    grid_2.append(row)


def check_duplicates(lst):
    for sub_list in lst:
        seen = set()
        for item in sub_list:
            if item != 0:
                if item in seen:
                    return True
                seen.add(item)
    return False


grid3 = []
for i in range(0, len(grid), 3):
    temp_lists = [[], [], []]
    for j in range(i, i + 3):
        sublist = grid[j]

        for k in range(3):
            start_index = k * 3
            end_index = start_index + 3
            temp_lists[k] += sublist[start_index:end_index]

    grid3.extend(temp_lists)


def dis(a):
    print("The solution is!")
    for i in range(z):
        for j in range(z):
            print(a[i][j], end="   ")
        print("\n")
        print()


def check(a, l):
    zero = 1
    for i in range(z):
        for j in range(z):
            if a[i][j] == 0:
                l[0] = i
                l[1] = j
                return True
    return False


def correct(grids, rows, col, num):
    # Check if 'num' is not in the current row, column, or 3x3 box
    for i in range(9):
        if (grids[rows][i] == num) or (grids[i][col] == num) or (grids[rows - rows % 3 + i // 3][col - col % 3 + i % 3]
                                                                 == num):
            return False
    return True


def sudoku(grids):
    pairs = [0, 0]
    if not check(grids, pairs):
        return True
    rows = pairs[0]
    cols = pairs[1]

    for num in range(1, 10):
        if correct(grid, rows, cols, num):
            grid[rows][cols] = num
            if sudoku(grid):
                return True

            grid[rows][cols] = 0
    return False


if check_duplicates(grid) or check_duplicates(grid_2) or check_duplicates(grid3):
    print("The puzzle cannot be solved!!\nThe Question is wrong!")
elif sudoku(grid):
    dis(grid)

File: sudoku.cpp
#include <iostream>
#include <fstream>
#include<math.h>
#include <vector>
#include <set>
//using namespace std;
#define z 9

bool dis(std::vector<std::vector<int>> & a) {
    std::cout << " The solution is! - " << std::endl;
    for (int i = 0; i < z; i++) {
        for (int j = 0; j < z; j++) {
            std::cout << a[i][j] << "   ";
        }
        std::cout << std::endl << std::endl;
    }
}


bool checkduplicates(const std::vector<std::vector<int>>& lst) {
    for (const auto& subList : lst) {
        std::set<int> seen;
        for (int item : subList) {
            if (item != 0) {
                if (seen.count(item) > 0) {
                    return true;
                }
                seen.insert(item);
            }
        }
    }
    return false;
}


bool check(std::vector<std::vector<int>>& a, std::vector<int>& pair) {
    for (int row = 0; row < z; row++) {
        for (int col = 0; col < z; col++) {
            if (a[row][col] == 0) {
                pair[0] = row;
                pair[1] = col;
                return true;
            }
        }
    }
    return false;
}

bool correct(std::vector<std::vector<int>>& grid, int row, int col, int num) {
    for (int i = 0; i < 9; i++) {
        if (grid[row][i] == num || grid[i][col] == num ||
            grid[row - row % 3 + i / 3][col - col % 3 + i % 3] == num) {
            return false;
        }
    }
    return true;
}

bool sudoku(std::vector<std::vector<int>>& grid) {
    std::vector<int> pairs(2, 0);
    if (!check(grid, pairs)) {
        return true;
    }
    int rows = pairs[0];
    int cols = pairs[1];

    for (int num = 1; num <= 9; num++) {
        if (correct(grid, rows, cols, num)) {
            grid[rows][cols] = num;
            if (sudoku(grid)) {
                return true;
            }

            grid[rows][cols] = 0;
        }
    }
    return false;
}

int main() 
{
	
    std::vector<std::vector<int>> grid(z, std::vector<int>(z, 0));

    std::ifstream file("sudoku.txt");
    if (file.is_open()) {
        int c = 0;
        for (int i = 0; i < z; i++) {
            std::vector<int> row(z, 0);
            for (int j = 0; j < z; j++) {
                file >> row[j];
                grid[i][j] = row[j];
                if (grid[i][j] == 0) {
                    c++;
                }
            }
        }
        file.close();
    }
    
    
    std::vector<std::vector<int>> grid2;
    
    for (int j = 0; j < 9; j++) {
        std::vector<int> row;
        for (int i = 0; i < 9; i++) {
            row.push_back(grid[j][i]);
        }
        grid2.push_back(row);
    }

    
    std::vector<std::vector<int>> grid3;

    for (int i = 0; i < grid.size(); i += 3) {
        std::vector<std::vector<int>> temp_lists = {{}, {}, {}};

        for (int j = i; j < i + 3; j++) {
            std::vector<int> sublist = grid[j];

            for (int k = 0; k < 3; k++) {
                int start_index = k * 3;
                int end_index = start_index + 3;
                temp_lists[k].insert(temp_lists[k].end(), sublist.begin() + start_index, sublist.begin() + end_index);
            }
        }

        grid3.insert(grid3.end(), temp_lists.begin(), temp_lists.end());
    }


    if (checkduplicates(grid) || checkduplicates(grid2) || checkduplicates(grid3)) {
        std::cout << "The puzzle cannot be solved!!\nThe Question is wrong!" << std::endl;
    }
	if (sudoku(grid)) {
        dis(grid);
    } else {
        std::cout << " The puzzle cannot be solved!!\nThe Question is wrong!" << std::endl;
    }

    return 0;
}
